app/components/candidate_interface/degree_required_component.rb

```diff
@@ -0,0 +1,75 @@
+class CandidateInterface::DegreeRequiredComponent < ViewComponent::Base
+  CANDIDATE_GRADES = {
+    'Upper second-class honours (2:1)' => 4,
+    'Lower second-class honours (2:2)' => 3,
+    'Third-class honours' => 2,
+    'Pass' => 1,
+  }.freeze
+
+  COURSE_REQUIRED_GRADES = {
+    'two_one' => 4,
+    'two_two' => 3,
+    'third_class' => 2,
+  }.freeze
+
+  COURSE_REQUIRED_GRADE_TEXT = {
+    'two_one' => '2:1 degree or higher (or equivalent)',
+    'two_two' => '2:2 degree or higher (or equivalent)',
+    'third_class' => 'Third-class degree or higher (or equivalent)',
+    'not_required' => 'Any degree grade',
+  }.freeze
+
+  DEGREE_GRADE_INSUFFICIENT_TEXT = {
+    'Lower second-class honours (2:2)' => 'a 2:2 degree.',
+    'Third-class honours' => 'a Third class degree.',
+    'Pass' => 'an Ordinary degree (pass).',
+  }.freeze
+
+  def initialize(application_choice)
+    @application_choice = application_choice
+    @course_degree_requirements = application_choice.current_course.degree_grade
+    @candidate_degrees = candidate_non_other_uk_degrees(application_choice)
+    @highest_candidate_degree_grade = ''
+  end
+
+  def candidate_degree_grade_below_required_grade?
+    return false if @candidate_degrees.empty? || degrees_with_other_grades_only(@candidate_degrees)
+
+    find_highest_degree_grade(@candidate_degrees)
+
+    COURSE_REQUIRED_GRADES[@course_degree_requirements] > CANDIDATE_GRADES[@highest_candidate_degree_grade]
+  end
+
+private
+
+  def find_highest_degree_grade(candidate_degrees)
```

Comment:
so i found this a bit tricky to parse what was going on, but dear lord it's tricky. Just spent about an hour and got to this point. Feel free to use what you like out of it 

``` 
 attr_accessor :application_choice, :course_degree_requirements, :candidate_degrees

  def initialize(application_choice)
    @application_choice = application_choice
    @course_degree_requirements = application_choice.current_course.degree_grade
    @candidate_degrees = candidate_non_other_uk_degrees(application_choice)
  end

  def candidate_degree_grade_below_required_grade?
    return false if @candidate_degrees.empty? || degrees_with_other_grades_only(@candidate_degrees)

    COURSE_REQUIRED_GRADES[@course_degree_requirements] > CANDIDATE_GRADES[highest_degree_grade]
  end

private

  def highest_degree_grade
    @highest_degree_grade ||= find_highest_degree_grade
  end

  def find_highest_degree_grade
    undergraduate_degree_grades = candidate_degrees.map(&:grade)
    CANDIDATE_GRADES.keys.each { |grade| return grade if undergraduate_degree_grades.include?(grade) }
  end

  def degrees_with_other_grades_only
    grades = candidate_degrees.map(&:grade)
    grades.any? { |grade| CANDIDATE_GRADES.keys.include?(degree.grade) }
  end
```

Reply:
then i just reused the method in the view 
![image](https://user-images.githubusercontent.com/42515961/132525542-579e57b9-0431-4389-ad6e-f48a4e950051.png)

Reply:
I had a go at trying to get the logic right (in my head), and this was where I got to, in pseudo-code.

Not sure if this is helpful or not, but basically I think the condition to show the warning is something like:

> Candidate has at least one UK undergraduate degree (bachelors) AND zero other degrees (non-UK or postgrads) AND their highest UK undergraduate degree grade is lower than the course requirement.

So if you've got a UK Masters or PhD, or any non-UK degree, then we don't show the warning (because we can't be sure if that would meet the requirements or not).

```ruby
def initialize(application_choice)
  @application_choice = application_choice
  @course_degree_requirements = application_choice.current_course.degree_grade
end

def show_degree_doesnt_meet_requirement_warning?
  candidate_only_has_uk_undergraduate_degrees &&
  candidate_highest_uk_undergraduate_degree < @course_degree_requirements
end

private

  def candidate_only_has_uk_undergraduate_degrees
    candidate_uk_undergraduate_degrees.size > 0 &&
    candidate_other_degrees.size == 0
  end
  
  def candidate_highest_uk_undergraduate_degree
    candidate_uk_undergraduate_degrees.sort {|degree| CANDIDATE_GRADES[degree.grade] }.last.grade
  end

  def candidate_uk_undergraduate_degrees
    application_choice.candidate.degrees.where(qualification_type_hesa_code: uk_undergraduate_degree_hesa_codes)
  end

  def uk_undergraduate_degree_hesa_codes
    HESA_DEGREE_TYPES.select {|degree_type| degree_type[3] == :bachelor }.collect(&:first)
  end
end
```

Reply:
Have just managed something similar will pysh up in a bit when the tests are passing.